### PR TITLE
feat: enumerative search of Substance mutations

### PR DESCRIPTION
**Pull request by wodeni**
# Description

Related issue/PR: #292, #601 

This PR refactors the `synthesis/Synthesizer` module in `@penrose/core` and includes some early experiments on inferring Substance program changes.

In #601,  we started encoding Substance program mutations (`Mutation`) as data. However, the program mutator still generates new programs by imperatively making randomized choices (e.g. picking which statement to mutate). This PR further allows the program mutator to produce new Substance programs by (1) enumerating all possible mutations per step and (2) randomly picking from a list of available `Mutation[]`. In this way, the logic of the program mutator is drastically simplified: to mutate a Substance program, the mutator simply executes (1) and (2) for `n` times: https://github.com/penrose/penrose/blob/dc86074d1053ed3768f8ed4bcc60ec2a38f71c76/packages/core/src/synthesis/Synthesizer.ts#L404

Another reason for this change is that we would like to infer program mutations from source and target programs. This will be very useful for future Edgeworth projects that involve programming-by-example/demonstration and/or templating. The new `synthesis/Search` module includes a set of functions for inferring mutations from source programs. Notably, `enumerateMutationPaths` is a lightly optimized version that finds all possible paths (up to some depth of there are cycles) between two Substance programs. The main optimizations are:

* Prune the environment to include only types appearing in the source and target program, not everything included in Domain
* “observational equivalence”: at each step, if multiple mutation paths lead to the same resulting program, we arbitrarily pick one and prune all others

https://github.com/penrose/penrose/blob/dc86074d1053ed3768f8ed4bcc60ec2a38f71c76/packages/core/src/synthesis/Search.ts#L473

# Implementation strategy and design decisions

Include a high-level summary of the implementation strategy and list important design decisions made, if any.

# Examples with steps to reproduce them

The program mutator doesn't have test suites, but the mutator search module does. Here's one test from `Search.test.ts`, for example. 

Goal: find all possible mutation paths between the following two Substance programs:

https://github.com/penrose/penrose/blob/4fc0fc0f5744fef83a4e4969bc9f67d71cb42e80/packages/core/src/synthesis/Search.test.ts#L301-L310

A reasonable guess is that `IsSubset(A, B)` is replaced by `Equal` with arguments swapped, and `Intersecting(A, B)` is refactored to `C := Intersection(A, B)`. By calling `enumerateMutationPaths` up to `depth=5`, the result looks like the following, and the shortest path is indeed the correct answer. 

```
@penrose/core:       [
@penrose/core:         [
@penrose/core:           'Set A\nSet B\nSet C\nEqual(B, A)\nIntersecting(A, B)',
@penrose/core:           'Replace the name of IsSubset(A, B) with Equal\n' +
@penrose/core:             'Swap arguments 0 and 1 of Equal(A, B)\n' +
@penrose/core:             'Change C := Intersection(A, B) to Intersecting(A, B)'
@penrose/core:         ],
@penrose/core:         [
@penrose/core:           'Set A\nSet B\nSet C\nEqual(B, A)\nIntersecting(A, B)',
@penrose/core:           'Replace the name of IsSubset(A, B) with Intersecting\n' +
@penrose/core:             'Replace the name of Intersecting(A, B) with Equal\n' +
@penrose/core:             'Swap arguments 0 and 1 of Equal(A, B)\n' +
@penrose/core:             'Change C := Intersection(A, B) to Intersecting(A, B)'
@penrose/core:         ],
@penrose/core:         [
@penrose/core:           'Set A\nSet B\nSet C\nEqual(B, A)\nIntersecting(A, B)',
@penrose/core:           'Replace the name of IsSubset(A, B) with Intersecting\n' +
@penrose/core:             'Replace the name of Intersecting(A, B) with IsSubset\n' +
@penrose/core:             'Replace the name of IsSubset(A, B) with Equal\n' +
@penrose/core:             'Swap arguments 0 and 1 of Equal(A, B)\n' +
@penrose/core:             'Change C := Intersection(A, B) to Intersecting(A, B)'
@penrose/core:         ]
@penrose/core:       ]
```

Performance wise, the search is enumerative, but the heuristics make it much better: `depth=10`, the naive search algorithm doesn’t terminate (I waited for 500s+) and the new one terminates in under 10s.

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions

* In `Synthesizer`, the configuration may include types that aren't present in the original AST, and the search for possible mutations will fail silently. In general, We don't keep track of statements that are added/updated by the synthesizer in the AST vs. original statements. Will tracking them give us better results?
* Seems like the Style program in `synthesizer-ui` is out-of-sync with the one in `examples`. I replaced it with the one from `examples`. Looks like `Collinear(A, E, C)` is not held in the example. Not super sure why (cc: @hsharriman):  
image

### CODE REVIEW

File: packages/core/src/synthesis/Search.ts [hunk 1]
```diff
@@ -350,54 +368,230 @@ export const findDiffs = (stmt: SubStmt<A>, diffs: StmtDiff[]): rdiffResult[] =>
...
-//     case "Delete":
-//     // case "Swap":
-//     // case "ReplaceName":
-//     case "TypeChange":
-//       return mutation.stmt;
-//   }
-// };
+//#region Single-mutation search
+
+export const cartesianProduct = <T>(...sets: T[][]) =>
```
Comment (samestep) [inline packages/core/src/synthesis/Search.ts:373]: ESLint warning fix:
```suggestion
export const cartesianProduct = <T>(...sets: T[][]): T[][] =>
```

**Review by samestep — APPROVED** [review]
LGTM! Gonna come to your office to chat a bit more to understand better what's going on :)

**Comment by samestep** [comment]
I tried running `yarn set-example` from `packages/synthesizer/` just now and it generated a bunch of files under `packages/automator/progs/`... perhaps we should put this line in `packages/automator/.gitignore`?
```gitignore
/progs/
```

File: packages/core/src/synthesis/Synthesizer.ts [hunk 2]
```diff
@@ -119,8 +121,15 @@ interface IDList {
...
 }
 
-export const initContext = (env: Env): SynthesisContext => {
+export const initContext = (
+  env: Env,
+  argOption: ArgOption,
+  argReuse: ArgReuse
+): SynthesisContext => {
   const ctx: SynthesisContext = {
+    generatedNames: Map<string, number>(),
```
[the remark this answers is not in the document]
Comment (wodeni) [reply packages/core/src/synthesis/Synthesizer.ts:130]: I think for vanilla JS `Map` you need the `new`. The one from `immutable` doesn't need it.

File: packages/core/src/compiler/Substance.ts [hunk 3]
```diff
@@ -84,6 +84,8 @@ export const compileSubstance = (
   const astOk = parseSubstance(prog);
   if (astOk.isOk()) {
     const ast = astOk.value;
+    // disambiguate statements in the AST
+    disambiguateFunctions(env, ast);
```
[the remark this answers is not in the document]
Comment (wodeni) [reply packages/core/src/compiler/Substance.ts:88]: I think it's only meant to be This is a bigger issue. The function mutates the AST :(, and the Substance checker could have done this much earlier in `checkSubstance`. I'm working on another branch to address this one.

[the remark this answers is not in the document]
Comment (wodeni) [reply packages/core/src/compiler/Substance.ts:88]: Resolved by #930

**Pull request merged**